### DISCUSSION TITLE
Sobre o MeuPluggy

### DISCUSSION THREAD

**Discussion by diegomdrs** [andreroggeri/pynubank#431] [General]
Fala ae pessoal.

Estou desenvolvendo uma integração ao MeuPluggy (Open Finance) como uma alternativa a um possível fim do acesso ***simplificado*** à API de cartões de crédito do Nubank, assim como ocorreu recentemente com a API da conta corrente.

Descobri sobre o pluggy.ia/MeuPluggy aqui mesmo na comunidade em alguma thread aleatória e fiquei empolgado com a ideia, pois mesmo não tendo a mesma frequência de atualização da API nativa do Nubank, tornaria possível obter os dados diariamente. E melhor: tornaria possível obter os dados da conta e cartões de outros bancos, como BB, Caixa e etc. Pra facilitar, os caras do projeto disponibilizaram uma conta dev free pro serviço, com algumas limitações mas ainda assim muito útil.

Tudo muito empolgante, mas no dia a dia não tem funcionado. Na minha aplicação tenho tanto a API do Nubank quanto a integração ao MeuPluggy para cartões de crédito. Dessa forma, consigo comparar diretamente o resultado das duas formas de acesso e seus atributos equivalentes. Infelizmente, os resultados são muito diferentes em vários aspectos, dentre eles:

- Algumas transações só são retornadas nas primeiras chamadas após a primeira autenticação ao MeuPluggy. Após outras chamadas com os mesmos parâmetros, token atualizado, solicitações de atualização de um Item (uma entidade parte da API), status corretos, algumas transações não são mais retornadas

- Transações com valores negativos como adiantamentos de parcelas e adiantamento de fatura não são retornados. Transações como PIX no crédito também não são retornadas.

Na comunidade no Discord do Pluggy, alguns desenvolvedores do projeto até tentam ajudar mas tenho receio que, na real, o negócio tenha algumas limitações no que pode ou não ser retornado diferentemente do que estamos acostumados por aqui.

Aí pergunto a quem está utilizando essa solução aqui na comunidade, obtendo as transações do cartão de crédito: essa integração está funcionando pra vocês de verdade mesmo? Ou melhor: tem alguém usando por aqui? Como tem sido a experiência? Na conta corrente e outros serviços do MeuPluggy através do Open Finance, é diferente?

Valeu!

**Comment by felipythondev** [comment]
Pô eu cheguei a criar uma conta para usar mas não cheguei a implementar a integração. Queria uma possibilidade de ver dados de tudo que tem na minha conta. Cartão, Conta, investimentos também. Se conseguir pegar de outros bancos também vai ser perfeito.

**Comment by diegomdrs** [comment]
Em teoria, é exatamente isso que o MeuPluggy disponibiliza, tanto na conta dev free como no plano pago.

Porém, na vida real, ao menos no cartão de crédito, essas transações que citei não são retornadas. Aí meio que invalida a integração uma vez que você não pode confiar na resposta do serviço.

**Comment by hugobarauna** [comment]
Estou usando para puxar dados de conta corrente da Nubank, porém não retorna os dados tão bem quanto o https://github.com/andreroggeri/pynubank retornava.

Não vem o nome da pessoa/empresa para quem paguei. Na prática tenho conseguido puxar apenas data e valor (R$) de cada transação da nuconta. Eu puxo isso e sincronizo com meu [YNAB](https://www.ynab.com/), via API deles, daí depois manualmente vou colocando o nome da empresa/pessoa de cada transação no meu YNAB.

**Reply by tavlima** [reply]
Uma coisa que vi no Pluggy é o CPF/CNPJ. Talvez vc possa usar isso para mapear e traduzir até mesmo antes de enviar pro YNAB.

A propósito, estou deixando o YNAB de lado ($$$) e adotando o ActualBudget, self-hosted. Se uma galera daqui se interessasse em contribuir pro projeto, seria massa. Falta tradução pra pt-br, por exemplo.

**Comment by tavlima** [comment]
Eu tenho usado e realmente era melhor pelo pynubank. Mas o pessoal do Pluggy é bem atencioso e tentam resolver as limitações, na medida que os dados são expostos pela API OpenFinance do Nubank. Essa questão de não ter as anotações no PIX, por exemplo, é bem chato.

Estou usando pra Nubank, XP e Caixa, atualmente. Deu algum xabu na do BB, mas é a próxima.

Eu sugiro dar um pulo lá no Discord deles e reportar os bugs. É bom pro negócio deles e pra gente, que consegue o que precisa. Pinga o @gabrielpanga, se precisar.

**Reply by Gabrielpanga** [reply]
Obrigado @tavlima  pela menção!

Claro que no canal discord você pode enviar qualquer problema, e nós faremos acompanhamento para resolver.
Muitas vezes dependemos das instituições, mas mesmo que demore alguns dias, elas acabam dando uma resposta!

**Comment by diegomdrs** [comment]
> Eu tenho usado e realmente era melhor pelo pynubank. Mas o pessoal do Pluggy é bem atencioso e tentam resolver as limitações, na medida que os dados são expostos pela API OpenFinance do Nubank. Essa questão de não ter as anotações no PIX, por exemplo, é bem chato.
> 
> Estou usando pra Nubank, XP e Caixa, atualmente. Deu algum xabu na do BB, mas é a próxima.
> 
> Eu sugiro dar um pulo lá no Discord deles e reportar os bugs. É bom pro negócio deles e pra gente, que consegue o que precisa. Pinga o @Gabrielpanga, se precisar.

Muito bom. É com ele mesmo que converso frequentemente. Quando encontro algum erro, dou o feedback na comunidade e principalmente pro Gabriel na DM do Discord. O cara é gente fina. Inclusive, alguns problemas que tive (e outros usuários provavelmente), eles abriram um ticket para o Nubank e foi resolvido. Imagino que dependendo da instituição eles ficam de mãos atadas e só podem contar com a boa vontade da galera.

A sua solução é para obter as transações do cartão de crédito também? Caso seja, como lida com as transações que faltam? Para algumas eu criei um cache para ir acumulando, com base no id da transação. Dessa forma eu consigo meio que contornar as transações que são retornadas e as que repentinamente somem da resposta do GET /transactions.

**Reply by tavlima** [reply]
Cara, não tive problema de transação faltando. O que vi foram algumas transações com sinal errado, mas outro usuário reportou e já corrigiram. Mas uma coisa que é um caos são transações parceladas: os metadados são pouco consistentes, datas não coincidem, etc. Fiz uma lógica onde só considero a primeira parcela que vem do Pluggy e descarto as demais. Aí crio as outras “na mão”, o que acaba incorrendo num valor errado de parcela, dados os ajustes que o Nubank faz quando a divisão exata não é possível. Como as vezes faltam os metadados de valor total da compra, fica impossível de resolver no código.

**Reply by Gabrielpanga** [reply]
As transações parceladas são um desafio. Atualmente disponibilizamos os metadados que em qualquer caso se não estiverem incluídos ou estiverem incompletos, você pode nos reportar para enviar ao nubank como erro.

Em dezembro, o Nubank demorou um mês para corrigir problemas de inconsistência de metadados, o que fez com que diversas transações ficassem incorretas.

O campo creditCardMetadata é obrigatório para que as instituições reportem corretamente.

Comente também que as transações parceladas são recuperadas na primeira compra, já constando na sua data futura.
